cli/src/codecs.js

```diff
@@ -0,0 +1,363 @@
+            input_height,
+            width,
+            height,
+            resizeNameToIndex(method),
+            premultiply,
+            linearRGB,
+          ),
+          width,
+          height,
+        );
+      };
+    },
+    defaultOptions: {
+      method: 'lanczos3',
+      fitMethod: 'stretch',
+      premultiply: true,
+      linearRGB: true,
+    },
+  },
+  // // TODO: Need to handle SVGs and HQX
+  quant: {
+    name: 'ImageQuant',
+    description: 'Reduce the number of colors used (aka. paletting)',
+    instantiate: async () => {
+      const imageQuant = await imageQuantPromise;
+      return (buffer, width, height, { numColors, dither }) =>
+        new ImageData(
+          imageQuant.quantize(buffer, width, height, numColors, dither),
+          width,
+          height,
+        );
+    },
+    defaultOptions: {
+      numColors: 255,
+      dither: 1.0,
+    },
+  },
+  rotate: {
+    name: 'Rotate',
+    description: 'Rotate image',
+    instantiate: async () => {
+      return async (buffer, width, height, { numRotations }) => {
+        const degrees = (numRotations * 90) % 360;
+        const sameDimensions = degrees == 0 || degrees == 180;
+        const size = width * height * 4;
+        const { instance } = await WebAssembly.instantiate(
+          await fsp.readFile(pathify(rotateWasm)),
+        );
+        const { memory } = instance.exports;
+        const additionalPagesNeeded = Math.ceil(
+          (size * 2 - memory.buffer.byteLength + 8) / (64 * 1024),
+        );
+        if (additionalPagesNeeded > 0) {
+          memory.grow(additionalPagesNeeded);
+        }
+        const view = new Uint8ClampedArray(memory.buffer);
+        view.set(buffer, 8);
+        instance.exports.rotate(width, height, degrees);
+        return new ImageData(
+          new Uint8ClampedArray(view.slice(size + 8, size * 2 + 8)),
+          sameDimensions ? width : height,
+          sameDimensions ? height : width,
+        );
+      };
+    },
+    defaultOptions: {
+      numRotations: 0,
+    },
+  },
+};
+
+export const codecs = {
+  // mozjpeg: {
+  //   name: "MozJPEG",
+  //   extension: "jpg",
+  //   detectors: [/^\xFF\xD8\xFF/],
+  //   dec: () => instantiateEmscriptenWasm(mozDec, mozDecWasm),
+  //   enc: () => instantiateEmscriptenWasm(mozEnc, mozEncWasm),
+  //   defaultEncoderOptions: {
+  //     quality: 75,
+  //     baseline: false,
+  //     arithmetic: false,
+  //     progressive: true,
+  //     optimize_coding: true,
+  //     smoothing: 0,
+  //     color_space: 3 /*YCbCr*/,
+  //     quant_table: 3,
+  //     trellis_multipass: false,
+  //     trellis_opt_zero: false,
+  //     trellis_opt_table: false,
+  //     trellis_loops: 1,
+  //     auto_subsample: true,
+  //     chroma_subsample: 2,
+  //     separate_chroma_quality: false,
+  //     chroma_quality: 75
+  //   },
+  //   autoOptimize: {
+  //     option: "quality",
+  //     min: 0,
+  //     max: 100
+  //   }
+  // },
+  webp: {
```

Comment:
All of this (extensions, default options) also feels like something we had largely in `encode_meta` modules - didn't we? Can we reuse/extend them?

Reply:
Yeah, this is one of these duplications that is going to be much easier to remove once we land this. Also the main project is TypeScript while the CLI currently is not. That should probably get changed as well at some point.